### DISCUSSION TITLE
Dll From Web Worker

### DISCUSSION THREAD

**Discussion by acransac** [webpack/webpack#13237] [Q&A]
Hi,

I have trouble using a dll (I mean the output of a webpack build with the DllPlugin) from a web worker. I put up a repository with the simplest setup I could figure out that reproduces the error: https://github.com/acransac/issue-webpack-dll-from-web-worker.

In summary: I have two entries "dll.js" and "app.js". I build the first one with the DllPlugin in a dedicated configuration. The second entry uses the dll (via DllReferencePlugin) in the main thread and starts a web worker that uses the same dll. The dll export is found in the main thread but from the worker I get the "dll is not defined" error.

Would someone take a look at my setup in the above repository and point me to configuration mistakes or possible work-arounds?

Thanks for your help, should you need to know something more I will answer.

PS: I use Webpack 5.35.1

**Comment by alexander-akait** [comment]
Yes, there is the problem what `Worker`s have own scope...
/cc @sokra

**Reply by acransac** [reply]
Thanks for the answer @alexander-akait . Is there anything I can do with the config so that the dll asset is available within the worker's scope?

**Reply by alexander-akait** [reply]
In theory you can create two dlls, and for second use `sourceType: 'self'`, it is impossible to share something between window and worker scope

**Reply by acransac** [reply]
Do you mean `library.type = "self"` in the dll's Webpack configuration?

I tried this and it doesn't seem to work, I get the same error ("dllworker is not defined").

To setup that approach I added a second dll output with `library.type = "self"`, then separated the worker in the app's Webpack configuration (not using Webpack's built-in worker loader) so I could use DllReferencePlugin with the second dll's manifest. I updated the HTML to download the two dlls but the result is eventually the same. You can find the exact setup I used in the repository I mentioned above in the branch "library-type-self".

**Reply by alexander-akait** [reply]
Is there anything else I can help?

**Reply by acransac** [reply]
Just to make sure it is clear, the `library.type = self` approach **did not** work. The second dll is still not found from the worker.

I see in Chrome's DevTools Sources tab that both dlls are not in the worker environment:
![image](https://user-images.githubusercontent.com/34621976/117717147-ba35b180-b1da-11eb-8057-df7dcf04dc46.png)

**Reply by alexander-akait** [reply]
You need import them on worker code side

**Reply by alexander-akait** [reply]
Maybe you can share current solution?

**Reply by acransac** [reply]
I did import the dll on the worker side (with the standard `import ... from ... `, and I also tried with `importScripts`).

As I mentioned before, you can find the current solution here:  https://github.com/acransac/issue-webpack-dll-from-web-worker . The original trial is in the master branch, the one with a second dll dedicated to the worker (with `library.type = self`) is in the branch "library-type-self".
